bitcoin/core/key.py

```diff
@@ -35,15 +35,142 @@
+_ssl.EC_KEY_get0_group.argtypes = [ctypes.c_void_p]
+
+_ssl.EC_KEY_get0_public_key.restype = ctypes.c_void_p
+_ssl.EC_KEY_get0_public_key.argtypes = [ctypes.c_void_p]
+
+_ssl.EC_KEY_set_conv_form.restype = None
+_ssl.EC_KEY_set_conv_form.argtypes = [ctypes.c_void_p, ctypes.c_int]
+
+_ssl.EC_KEY_set_private_key.restype = ctypes.c_int
+_ssl.EC_KEY_set_private_key.argtypes = [ctypes.c_void_p, ctypes.c_void_p]
+
+_ssl.EC_KEY_set_public_key.restype = ctypes.c_int
+_ssl.EC_KEY_set_public_key.argtypes = [ctypes.c_void_p, ctypes.c_void_p]
+
+_ssl.EC_POINT_free.restype = None
+_ssl.EC_POINT_free.argtypes = [ctypes.c_void_p]
+
+_ssl.EC_POINT_is_at_infinity.restype = ctypes.c_int
+_ssl.EC_POINT_is_at_infinity.argtypes = [ctypes.c_void_p, ctypes.c_void_p]
+
+_ssl.EC_POINT_new.restype = ctypes.c_void_p
+_ssl.EC_POINT_new.argtypes = [ctypes.c_void_p]
+
+_ssl.EC_POINT_mul.restype = ctypes.c_int
+_ssl.EC_POINT_mul.argtypes = [ctypes.c_void_p, ctypes.c_void_p, ctypes.c_void_p, ctypes.c_void_p, ctypes.c_void_p, ctypes.c_void_p]
+
+_ssl.EC_POINT_set_compressed_coordinates_GFp.restype = ctypes.c_int
+_ssl.EC_POINT_set_compressed_coordinates_GFp.argtypes = [ctypes.c_void_p, ctypes.c_void_p, ctypes.c_void_p, ctypes.c_int, ctypes.c_void_p]
+
+_ssl.ECDSA_sign.restype = ctypes.c_int
+_ssl.ECDSA_sign.argtypes = [ctypes.c_int, ctypes.c_void_p, ctypes.c_int, ctypes.c_void_p, ctypes.c_void_p, ctypes.c_void_p]
+
+_ssl.ECDSA_size.restype = ctypes.c_int
+_ssl.ECDSA_size.argtypes = [ctypes.c_void_p]
+
+_ssl.ECDSA_verify.restype = ctypes.c_int
+_ssl.ECDSA_verify.argtypes = [ctypes.c_int, ctypes.c_void_p, ctypes.c_int, ctypes.c_void_p, ctypes.c_int, ctypes.c_void_p]
+
+_ssl.ECDSA_SIG_free.restype = None
+_ssl.ECDSA_SIG_free.argtypes = [ctypes.c_void_p]
+
+_ssl.ECDH_compute_key.restype = ctypes.c_int
+_ssl.ECDH_compute_key.argtypes = [ctypes.c_void_p, ctypes.c_int, ctypes.c_void_p, ctypes.c_void_p]
+
+_ssl.ERR_error_string_n.restype = None
+_ssl.ERR_error_string_n.argtypes = [ctypes.c_ulong, ctypes.c_char_p, ctypes.c_size_t]
+
+_ssl.ERR_get_error.restype = ctypes.c_ulong
+_ssl.ERR_get_error.argtypes = []
+
+_ssl.d2i_ECDSA_SIG.restype = ctypes.c_void_p
+_ssl.d2i_ECDSA_SIG.argtypes = [ctypes.c_void_p, ctypes.c_void_p, ctypes.c_long]
+
+_ssl.d2i_ECPrivateKey.restype = ctypes.c_void_p
+_ssl.d2i_ECPrivateKey.argtypes = [ctypes.c_void_p, ctypes.c_void_p, ctypes.c_long]
+
+_ssl.i2d_ECDSA_SIG.restype = ctypes.c_int
+_ssl.i2d_ECDSA_SIG.argtypes = [ctypes.c_void_p, ctypes.c_void_p]
+
+_ssl.i2d_ECPrivateKey.restype = ctypes.c_int
+_ssl.i2d_ECPrivateKey.argtypes = [ctypes.c_void_p, ctypes.c_void_p]
+
+_ssl.i2o_ECPublicKey.restype = ctypes.c_void_p
+_ssl.i2o_ECPublicKey.argtypes = [ctypes.c_void_p, ctypes.c_void_p]
+
+# Why do tests fail when uncommenting this? What other errors are being
+# masked (if any) by weird ctypes usages or misuse of ctypes.c_void_p?
+# _ssl.o2i_ECPublicKey.restype = ctypes.c_void_p
+# _ssl.o2i_ECPublicKey.argtypes = [ctypes.c_void_p, ctypes.c_void_p, ctypes.c_long]
```

Comment:
I found the bug. In `CPubKey.__new__`, we have the line `self.is_fullyvalid = _cec_key.set_pubkey(self) != 0`. Without these declarations, pointers are returned as integers, and can be compared to 0. With these declarations, pointers are almost always returned as integers, except when they are `NULL` (which used to be 0). Then Python translates them to `None`, and `None != 0`. Therefore `is_fullyvalid` gets set to `True`. This can be easily fixed by using `self.is_fullyvalid = bool(_cec_key.set_pubkey(self))`, which works whether 0 or `None` is returned. Tested on my computer, this fixes both test failures.